### DISCUSSION TITLE
I've started a complete rewrite

### DISCUSSION THREAD

**Discussion by cmacrae** [cmacrae/spacebar#73] [News]
Just a quick post in the interest of sharing, as I'll be taking a break from features/fixes on spacebar for a while.  
I started a re-write from scratch the other day and I have to say it's going very well.  

### Overview (braindump ahead) 🧠 
- Sticking with C so it's nice and fast. It's also been a great learning experience so far :)
- Will be calling it something different, but keeping the repository. I don't think the name `spacebar` was originally thought out too well - it makes it very hard for people to search for.
- ~For event driven actions, I'm hoping to potentially consume them from yabai. This decision is not set in stone, as it means this project will rely on yabai for these parts. There is a lot of code hanging around to support these features which are already implemented in yabai - it'd greatly simplify the project if it were to subscribe to events/await signals (using yabai's signalling system). This is the biggest TBD and not a light decision.~ I've looked at this some more and actually, it may not need to be external! 🎉 
- Modular approach. I've already built out the core composition - each piece of information is a "module" which represents a common set of attributes: a CoreGraphics window, text (optional), icon (optional), text font, icon font, refresh loop, colours, coordinates. This means each part of these can be treated individually, which grants a few nice things, such as different rate of refresh (for instance, the clock can be drawn every second whereas a mail indicator can be drawn every 5 seconds), different fonts per module, as many modules at any position as the user desires, etc.
- Hoping to implement mouse click actions per module that's configurable. It'll be needed for the coffee module (right now, if you have it active, it'll just inhibit sleep) to toggle state and lots of people have shown interest in being able to change spaces with a click on the indicator. I imagine an approach where each module has a custom action you can associate, a shell command, perhaps. As for the native modules, some of these would be predefined.
- With the above, maybe even a hover action, spawning little overlay windows might be a nice feature, but I may be getting ahead of myself.
- With the much cleaner/simpler modular composition, I hope to implement a more fluid configuration interface. The messy approach of a huge if/else tree to interpret user configuration will be a thing of the past, given the dynamic nature of the new modules. This could potentially allow users to express their bar as a YAML/JSON manifest. I do quite like the message/socket approach for some stuff, though - so I'm undecided and the configuration layer is still a ways off.
- A suite of "native modules" is provided. The following are already implemented: clock, power, shell output, coffee (sleep inhibitor).
- Highly commented code. Hardly a feature, but a welcome addition, I'm sure. I've found implementing new modules an absolute breeze so far and want to make that easy for others to do, where they find the shell output module doesn't meet their needs (hopefully this'll be a seldom circumstance!).
- Developing privately for now, but it'll all suddenly appear in this repo when it's ready, along with the rename. I'd like to get feature parity with the current implementation before doing that, of course.

Anyways, wish me luck! 🤞 It's off to a great start, so I hope I'll have something a lot nicer to share with you all soon :)

**Comment by meseck** [comment]
Thanks for the great work so far. I'm really looking forward to the new version!

I would be happy to help out with a logo if needed. Just for reference and to see if you like my style. I am the creator of the [yabai](https://github.com/koekeishiya/yabai) logo and also made one for [AceJump](https://github.com/acejump/AceJump).

Just let me know. I'm always happy to support a tool that I like and use every day! 👍

**Reply by cmacrae** [reply]
You're very welcome! Thank you for the support :) I can't wait to share what I've been working on

That's really kind of you to offer! Thank you so much 🙇 I love both the yabai & AceJump logos 👏   
Illustration (but certainly not graphic design) is a hobby of mine, so I'm planning on whipping something up myself once I've settled on a new name. But that certainly doesn't mean we couldn't work together! 🧑‍🎨

**Reply by meseck** [reply]
Thank you for your feedback! I'm always happy to help out. A collaboration sounds like fun. 😊
 
As I said, let me know if there is a need.

**Comment by corpooo** [comment]
Hi! 

First of all I want to say thank you for this awesome project! It's an awesome addition when used alongside yabai.

I am also curious how the rewrite is going. 😄

Great job and have a nice day!

**Reply by carsin** [reply]
If you haven't already, check out #84 and https://github.com/FelixKratz/SketchyBar

**Comment by davay** [comment]
@cmacrae Sorry I don't mean to bother but I'm wondering if a rewrite for this is still happening or if efforts are concentrated for SketchyBar?
